wcfsetup/install/files/lib/action/FileDownloadAction.class.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+namespace wcf\action;
+
+use Laminas\Diactoros\Response;
+use Laminas\Diactoros\Response\EmptyResponse;
+use Laminas\Diactoros\Stream;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\data\file\File;
+use wcf\http\ContentDisposition;
+use wcf\http\Helper;
+use wcf\system\exception\IllegalLinkException;
+use wcf\system\exception\PermissionDeniedException;
+use wcf\util\FileUtil;
+
+/**
+ * Offers a file for download.
+ *
+ * @author Alexander Ebert
+ * @copyright 2001-2024 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @since 6.1
+ */
+final class FileDownloadAction implements RequestHandlerInterface
+{
+    public function handle(ServerRequestInterface $request): ResponseInterface
+    {
+        $parameters = Helper::mapQueryParameters(
+            $request->getQueryParams(),
+            <<<'EOT'
+                array {
+                    id: positive-int
+                }
+                EOT,
+        );
+
+        $file = new File($parameters['id']);
+        if (!$file->fileID) {
+            throw new IllegalLinkException();
+        }
+
+        $processor = $file->getProcessor();
+        if ($processor === null) {
+            throw new IllegalLinkException();
+        }
+
+        if (!$processor->canDownload($file)) {
+            throw new PermissionDeniedException();
+        }
+
+        $eTag = \sprintf(
+            '"%d-%s"',
+            $file->fileID,
+            \substr($file->fileHash, 0, 8),
+        );
+
+        $httpIfNoneMatch = $_SERVER['HTTP_IF_NONE_MATCH'] ?? '';
+        if ($httpIfNoneMatch === $eTag) {
+            return new EmptyResponse(304);
```

Comment:
The 304 should include the caching headers (cache-control + expires): https://www.rfc-editor.org/rfc/rfc9110#name-304-not-modified. The easy solution might be including all of them by means of a helper function.